### ISSUE TITLE
security considerations

### ISSUE THREAD

**Issue by hasufell** [GaloisInc/haskell-tor#23]
I'm just throwing this out here for discussion since this seems to be one of the first real-world implementations of crucial security-related applications in haskell.

I've asked on haskell-cafe ML before: how much do we actually know about haskell to say whether it's safe to use for such a use case? What problems (e.g. explicit memory management) do we face here and need to be concerned about? Can haskell introduce additional attack vectors (e.g. like timing problems in the JVM)?

Also a few posts about this or similar topics:
https://mail.haskell.org/pipermail/haskell-cafe/2015-February/118059.html
http://www.leonmergen.com/haskell/crypto/2015/03/21/on-the-state-of-cryptography-in-haskell.html

**Comment by acw** [comment]
Well, let me counter: how much do you know about any language to say that it's safe for a security critical use case? We have nearly 40 years of evidence stating that C is a terrible choice, for example, but people still write security-critical software in it. You can make similar arguments for any language of your choosing.

In the end, if you want a completely trustworthy system, you need a complete and precise description of all the threats; a complete and precise description of the semantics of all the languages, tools, and complete software stack involved; and a complete and precise description of the security properties you wish to enforce. The only relevance your choice of language has is in your ability to describe the semantics of the language and how the underlying primitives make meeting your goals easier or harder.

In this particular case, I, personally, am comfortable offloading the responsibility for a lot of timing-level attacks to the author of the `cryptonite` library, which `haskell-tor` uses for its crypto. Otherwise, I believe that the safety provide by the strong type system overwhelms the possibility that `haskell-tor` is leaking information due to timing problems. And I'm pretty happy to leave it at that.

Obviously, I find these issues somewhat interesting, as the company I work for (which is the company under whose GitHub organization this project is hosted) thinks about these as part of our core competency. But for this project, I'm happy with the low- / medium-assurance guarantees that Haskell offers via its type system, and how we can best leverage them. After all, it doesn't run on anything approaching a verified OS, unless someone does a port of the HaLVM to seL4.

I'll leave this issue open, just in case others really want to chime in, but if anyone has any concrete suggestions, please file a separate issue.

**Comment by hasufell** [comment]
> We have nearly 40 years of evidence stating that C is a terrible choice, for example, but people still write security-critical software in it.

Indeed. But then again, we know pretty well by now _why_ it's a terrible choice and what can go wrong. I'm not really criticizing anyone for choosing haskell, I have actually considered doing some fun-crypto in haskell too (NTRUEncrypt to be exact). I guess what I'm really after is sort of a list of pitfalls that are more or less specific to the language (laziness?) and can affect security systems. And this project is sort of a pioneer, so maybe it makes sense to add a section to the Readme (or another file) that explains why haskell was chosen and also add any sort of security considerations that came up during development.

**Comment by hasufell** [comment]
Here's another thought that came up while studying security design patterns. In the report [Secure Design Patterns](https://www.sei.cmu.edu/reports/09tr010.pdf) by the [SEI](https://www.cert.org/), there's a section "4.2 **Clear Sensitive Information**". The question now arises with a garbage collected language... do we know that:
1. the garbage collection is not predictable in some way, and
2. it can't be compromised

If we don't trust the GC to always do the right thing, then we basically need means to explicitly clear sensitive information, also the one in memory. Which might be particularly problematic for a lazy functional language.

**Comment by hasufell** [comment]
I also found this issue interesting https://github.com/mitls/mitls-fstar/issues/124

Especially

> What we consider is indeed a programming discipline which ensures that the execution path is independent from the secret data.

So this goes into the direction of information flow policy, security type systems etc.
Interesting reads on this topic:
https://www.cis.upenn.edu/~stevez/papers/LZ06a.pdf
http://www.cse.chalmers.se/~russo/publications_files/haskell22-russo.pdf
